src/bosh_azure_cpi/lib/cloud/azure/vm_manager.rb

```diff
@@ -194,6 +224,12 @@ def delete(instance_id)
         end
       end
 
+      # After VM is deleted, tasks that can be run in parallel are:
+      #   * deleting empty availability set
```

Comment:
If CPI fails to delete availability set for some reason, while all the network interfaces are deleted. Then CPI can't delete the availability sets based on the tags when retrying. Is it possible?

Reply:
Logically, it is possible. In the case you have mentioned, it should be handled in deleting availability set, not in deleting NICs. I even recommend not tagging avset name in NIC, we were coupling two independent resources. 
And further more, did you see any error that failed in deleting avset?  Did we make the logic too heavy to handle such issues which is not very possible to happen?

Reply:
> did you see any error that failed in deleting avset?
I also don't have the answers. Maybe the telemetry data can help answer it.

https://github.com/cloudfoundry-incubator/bosh-azure-cpi-release/blob/bd9679e90106e2e644ce121c0ee18d89fe7e6fd3/src/bosh_azure_cpi/lib/cloud/azure/vm_manager.rb#L199-L208 handles the logic in current codes. I agree that it's a corner case. These logic's purpose is to handle cases which are not very possible to happen.
Whether it's worth to do so or not is another open question.